javascript/sbp/observation.js

```diff
@@ -0,0 +1,1016 @@
+  ----------
+  P : int
+    Pseudorange observation
+  L : CarrierPhase
+    Carrier phase observation
+  cn0 : int
+    Carrier-to-Noise density
+  lock : int
+    Lock indicator. This value changes whenever a satellite
+signal has lost and regained lock, indicating that the
+carrier phase ambiguity may have changed.
+
+  sid : int
+    Signal identifier of the satellite signal - values 0x00
+through 0x1F represent GPS PRNs 1 through 32 respectively
+(PRN-1 notation); other values reserved for future use.
+
+  sender : int
+    Optional sender ID, defaults to SENDER_ID (see sbp/msg.py).
+
+  */
+
+  this.sbp = sbp;
+  this.fields = this.parser.parse(msg);
+
+  return this;
+};
+PackedObsContent.prototype = Object.create(SBP.prototype);
+PackedObsContent.prototype.constructor = PackedObsContent;
+PackedObsContent.prototype.toBuffer = function () {
+  var buffers = [];
+  var b;
+  b = new Buffer( 4 );
+  b.writeUInt32LE(this.fields.P);
+  buffers.push(b);
+  buffers = buffers.concat( CarrierPhase.prototype.toBuffer.apply({ fields: this.fields.L }));
+  b = new Buffer( 1 );
+  b.writeUInt8(this.fields.cn0);
+  buffers.push(b);
+  b = new Buffer( 2 );
+  b.writeUInt16LE(this.fields.lock);
+  buffers.push(b);
+  b = new Buffer( 4 );
+  b.writeUInt32LE(this.fields.sid);
+  buffers.push(b);
+  return Buffer.concat(buffers);
+};
+PackedObsContent.prototype.parser = new Parser()
+  .endianess('little')
+  .uint32('P')
+  .nest('L', { type: CarrierPhase.prototype.parser })
+  .uint8('cn0')
+  .uint16('lock')
+  .uint32('sid');
+
+var MSG_OBS = function (sbp, msg) {
+  
+  /*SBP class for message MSG_OBS (0x0043).
+  
+
+  
+  The GPS observations message reports all the raw pseudorange and
+carrier phase observations for the satellites being tracked by
+the device. Carrier phase observation here is represented as a
+40-bit fixed point number with Q32.8 layout (i.e. 32-bits of
+whole cycles and 8-bits of fractional cycles).
+
+
+  Parameters
+  ----------
+  header : ObservationHeader
+    Header of a GPS observation message
+  obs : array
+    Pseudorange and carrier phase observation for a
+satellite being tracked.
+
+  sender : int
+    Optional sender ID, defaults to SENDER_ID (see sbp/msg.py).
+
+  */
+
+  this.sbp = sbp;
+  this.fields = this.parser.parse(msg);
+
+  return this;
+};
+MSG_OBS.prototype = Object.create(SBP.prototype);
+MSG_OBS.prototype.constructor = MSG_OBS;
+MSG_OBS.prototype.toBuffer = function () {
+  var buffers = [];
+  var b;
+  buffers = buffers.concat( ObservationHeader.prototype.toBuffer.apply({ fields: this.fields.header }));
+  for (var i in this.fields.obs) {
+    buffers = buffers.concat( PackedObsContent.prototype.toBuffer.apply({ fields: this.fields.obs[i] }));
+  }
+  return Buffer.concat(buffers);
+};
+MSG_OBS.prototype.parser = new Parser()
+  .endianess('little')
+  .nest('header', { type: ObservationHeader.prototype.parser })
+  .array('obs', { type: PackedObsContent.prototype.parser, readUntil: 'eof' });
```

Comment:
eof?

The observation messages can definitely be super-complicated to get right, from the declarative parsing point of view. Did this work? The tracking message is similar.